### PR TITLE
Alternative string-based approach to format string passthrough

### PR DESCRIPTION
**Pull request by joerick**
An alternative approach to, and closes #884 that uses regex instead of format_map. 

While working on #884 I felt that there were still quite a lot of gotchas in that implementation i.e.:

- curly brackets that look like positional replacement fields
- unmatched curly brackets

These would require the user to provide additional escaping a-la `{{` and `}}`.

This is an experiment to see if a non-format_map solution might work better.

I'm not sure which one I prefer! The main differences are that, in this PR,

- this approach doesn't have the gotchas mentioned above. Unrecognised syntax is always passed through.
- the escape sequences are different. This uses hash e.g. `#{wheel}` to passthrough a literal `{wheel}`, rather than `{{wheel}}`
- the escape sequence only works on variables known to the command. So in #884, doing `{{unknown}}` is passed through as `{unknown}`, but here the escape sequences are only removed on known params. e.g. `\{unknown}` passes through as `\{unknown}`, untouched. I think this is less disruptive, as most users won't have to think about the escaping unless they hit a naming conflict directly.
- If there are any `{{` escape sequences in the wild, due to the change in escaping rules, this would break that. However, I can only find [one such instance](https://sourcegraph.com/search?q=context:global+CIBW_%28BEFORE_%28ALL%7CBUILD%7CTEST%29%7C%28REPAIR_WHEEL%7CTEST%29_COMMAND%29.*%5B%5E%24%5D%7B%7B&patternType=regexp)

So I suppose I'm leaning towards this approach [^1]. What do you think @henryiii?

[^1]: Despite what I said in #840, sorry! Sometimes it takes writing the code to know.

### CODE REVIEW

**Comment by joerick** [comment]
Managed to fix the backslash issue. 

Having looked at it, `{{wheel}}` is actually quite tricky to ignore in regex, because you need to rely on both a negative lookbehind of `{` and a negative lookahead of `}` to properly match it. Perhaps `#{wheel}` would be okay as an escape sequence, what do you think?

**Comment by henryiii** [comment]
If you fixed the backslash, I would think that's fine. `#{}` looks like Ruby to be, but admittedly, it would probably be pretty rare for shell scripts, I think. No strong opinion here, I expect escaping this would be very uncommon.

**Comment by joerick** [comment]
Cool. I'm only really doing the escaping thing for completeness, I'd be surprised if it's ever used! I added a note about its use to the docs, too, all the way at the bottom of the Options page.

So this is ready for another review.

File: cibuildwheel/util.py [hunk 1]
```diff
@@ -48,14 +48,56 @@
...
+                }}  # literal close curly bracket
+            """,
+            re.VERBOSE,
+        )
+
+        # we use a lambda for repl to prevent re.sub interpreting backslashes
+        # in repl as escape sequences
+        result = re.sub(
+            pattern=find_pattern,
+            repl=lambda _: str(value),
```
Comment (henryiii) [inline cibuildwheel/util.py:83]:
```suggestion
            repl=value.replace('\\', r'\\'),
```

FYI, this is the suggestion in the Python docs for escaping here. Don't have a strong preference, just wanted to point it out. See the final suggestion in https://docs.python.org/3/library/re.html#re.escape.

**Review by henryiii — APPROVED** [review]

File: cibuildwheel/util.py (hunk 1 above)
Comment (joerick) [reply cibuildwheel/util.py:83]: Thanks, I missed that. Will stick to the lambda for convenience

**Pull request merged**